x-pack/plugins/security/server/authorization/validate_es_response.ts

```diff
@@ -13,48 +13,59 @@ export function validateEsPrivilegeResponse(
   actions: string[],
   resources: string[]
 ) {
-  const schema = buildValidationSchema(application, actions, resources);
-  const { error, value } = schema.validate(response);
-
-  if (error) {
-    throw new Error(
-      `Invalid response received from Elasticsearch has_privilege endpoint. ${error}`
-    );
+  const validationSchema = buildValidationSchema(application, actions, resources);
+  try {
+    validationSchema.validate(response);
+  } catch (e) {
+    throw new Error(`Invalid response received from Elasticsearch has_privilege endpoint. ${e}`);
   }
 
-  return value;
+  return response;
 }
 
 function buildActionsValidationSchema(actions: string[]) {
-  return Joi.object({
+  return schema.object({
     ...actions.reduce<Record<string, any>>((acc, action) => {
       return {
         ...acc,
-        [action]: Joi.bool().required(),
+        [action]: schema.boolean(),
       };
     }, {}),
-  }).required();
+  });
 }
 
 function buildValidationSchema(application: string, actions: string[], resources: string[]) {
   const actionValidationSchema = buildActionsValidationSchema(actions);
 
-  const resourceValidationSchema = Joi.object({
-    ...resources.reduce((acc, resource) => {
-      return {
-        ...acc,
-        [resource]: actionValidationSchema,
-      };
-    }, {}),
-  }).required();
+  const resourceValidationSchema = schema.object(
+    {},
+    {
+      unknowns: 'allow',
+      validate: (value) => {
+        const actualResources = Object.keys(value);
+        const unexpectedResource = actualResources.find((ar) => !resources.includes(ar));
+        if (unexpectedResource) {
+          throw new Error(`Unexpected resource in payload: ${unexpectedResource}`);
+        }
+        const missingResource = resources.find((r) => !actualResources.includes(r));
+        if (missingResource) {
+          throw new Error(`Missing expected resource in payload: ${missingResource}`);
+        }
```

Comment:
If I'm not mistaken, this has O(n^2) complexity and could be sped up quite a bit. I put together some quick benchmarks here: https://jsbench.me/ofkkpit5k2/1
* Current implementation: 9.28 ops/s
* Set comparison: 2241.66 ops/s
* Sorted array comparison: 4154.27 ops/s

Question: do we actually care to know whether a resource is unexpected or missing? If not, we could do something like this...

```suggestion
      validate: (value) => {
        const actualResources = Object.keys(value).sort();
        if (
          resources.length !== actualResources.length ||
          !resources.sort().every((x, i) => x === actualResources[i])
        ) {
          throw new Error('Payload did not match expected resources');
        }
```

Reply:
I debated simplifying the validation like you've described, but given the criticality of this interaction with ES, I'd prefer to keep this as-is.

That said, I like the perf gain of sorting the arrays ahead of time. I had considered that, but neglected to benchmark, so I'm glad you took the time to do so. I'll go ahead and make that change to see what kind of gains I can squeeze out of this.

Reply:
> I debated simplifying the validation like you've described, but given the criticality of this interaction with ES, I'd prefer to keep this as-is.

I understand. I'm not sure I agree, as with the current approach we could still be obscuring details -- if a response is missing resources _and_ contains unexpected resources, the error message only tells us one of those.

But if your concern is keeping the different error messages, we could also simplify it differently to retain those:

```ts
const actualResources = Object.keys(value).sort();
const expectedResources = resources.sort();

let missingResource: string | undefined;
let unexpectedResource: string | undefined;

for (let i = 0; i < actualResources.length; i++) {
  const actual = actualResources[i];
  const expected = expectedResources[i];
  if (actual !== expected) {
    if (expectedResources.includes(actual)) {
      missingResource = expected;
    } else {
      unexpectedResource = actual;
    }
    break;
  }
}

if (!missingResource && expectedResources.length > actualResources.length) {
  missingResource = expectedResources[actualResources.length];
}

if (unexpectedResource) {
  throw new Error(`Unexpected resource in payload: ${unexpectedResource}`);
}
if (missingResource) {
  throw new Error(`Missing expected resource in payload: ${missingResource}`);
}
```

Note that this snippet doesn't pass the unit tests as they are currently written, as it includes the sorting, and it detects the "missing expected resource" case before the "unexpected resource" case.

At any rate I think it's worth trying another approach, as O(n^2) complexity will get way worse with more than 3500 spaces.

Reply:
@jportner thanks for your persistence. I misread/misunderstood your original proposal. To answer your question for real this time, I don't really care if the resource is unexpected or missing -- I only care if there's a mismatch, but I don't need to report in which way the mismatch occurred.

Reply:
Done in [`56327de` (#90074)](https://github.com/elastic/kibana/pull/90074/commits/56327de8d96690552560e9739717100b2b17f5a2)